### PR TITLE
Update to modules-parent 1.13.7

### PR DESCRIPTION
**Pull request by bertfrees**
A lot of backward incompatible changes were made in the latest modules snapshots  (notably https://github.com/daisy/pipeline-modules/pull/5 and https://github.com/daisy/pipeline-modules/pull/23). This PR adapts to those changes, and at the same time does a big refactoring of the code, in the same spirit as https://github.com/daisy/pipeline-modules/pull/5. But most importantly it separates some (non NLB specific) functionality into separate modules, the goal being to incorporate these in Pipeline.

I initially just want to get your review and approval for these changes. I will then first incorporate the modules in Pipeline, after which the multi-module project can be turned into a simple project again, and then it can be merged.

There is one commit in particular for which I need your input: eedc88db51baab9e9ec3a80edd6e928f2fca8d65.

You should be able to build the project because I've uploaded all the snapshots to Sonatype.

### CODE REVIEW

**Comment by josteinaj** [comment]
Hi.

This sounds great. Sorry for not having reviewed yet. I've only looked at the commit you pointed out so far.

I'm guessing the change in eedc88d is fine, but I'd like to put a comment about it in the changelog for the next release. What change was made that resulted in these tests having to be updated?

**Comment by bertfrees** [comment]
The change was made in https://github.com/nlbdev/nordic-epub3-dtbook-migrator/pull/387/commits/f85ecb76feebffde0f5ed55acd4a77775eba5e35 and more specifically it has to do with these lines: https://github.com/nlbdev/nordic-epub3-dtbook-migrator/blob/f85ecb76feebffde0f5ed55acd4a77775eba5e35/nordic-epub3-dtbook-migrator/src/main/resources/xml/xslt/generic-to-nordic-dtbook.xsl#L65-L69

As far as I understand the problem is related to the interaction of two things:
- When `li` contains nested lists everything that is not a list is wrapped in `lic`.
- `a` are not allowed in Nordic DTBook and are therefore either unwrapped or replaced with `span`, depending on the context.

The former now happens in the generic XSLT (epub3-to-dtbook.xsl), the latter in generic-to-nordic-dtbook.xsl. Because of the lack of context, `a` inside `lic` is now always unwrapped. Apparently this didn't use to be the case when the `lic` was generated. So in a way the behavior is now more consistent.

File: nordic-epub3-dtbook-migrator/src/test/xprocspec/dtbook-to-html.step.xprocspec [hunk 1]
```diff
@@ -25,10 +29,10 @@
             <x:document type="port" port="in-memory.out"/>
         </x:context>
         <x:expect type="count" label="There should be exactly one output document" min="1" max="1"/>
-        <x:expect type="xpath" label="The output document (the intermediary HTML) should have an xml:base attribute with filename C00000.xhtml" test="tokenize(/*/@xml:base,'/')[last()]"
+        <x:expect type="xpath" label="The output document (the intermediary HTML) should have a base URI with filename C00000.xhtml" test="tokenize(/*/base-uri(.),'/')[last()]"
             equals="'C00000.xhtml'"/>
-        <x:expect type="xpath" label="The output document (the intermediary HTML) should have an xml:base attribute pointing to the temp-dir directory" test="/*/@xml:base"
-            equals="concat($temp-dir,'temp-dir/C00000.xhtml')"/>
+        <x:expect type="xpath" label="The output document (the intermediary HTML) should have a base URI pointing to the temp-dir directory" test="/*/base-uri(.)"
+            equals="concat($temp-dir,'temp-dir/C00000.xhtml')" pending="the base URI is normalized"/>
```
Comment (josteinaj) [inline nordic-epub3-dtbook-migrator/src/test/xprocspec/dtbook-to-html.step.xprocspec:35]: Does this test have to be pending? Do we need to change something to make this work?

File: nordic-epub3-dtbook-migrator/src/test/xprocspec/html-to-epub3.step.xprocspec [hunk 2]
```diff
@@ -71,10 +75,10 @@
                 <d:file href="EPUB/nav.ncx" method="xml" indent="true" omit-xml-declaration="false" media-type="application/x-dtbncx+xml"/>
             </x:document>
             <x:document type="inline" base-uri="temp-dir">
-                <d:file href="EPUB/package.opf" method="xml" indent="true" omit-xml-declaration="false" media-type="application/oebps-package+xml"/>
+                <d:file href="EPUB/nav.xhtml" method="xhtml" indent="true" omit-xml-declaration="false" media-type="application/xhtml+xml" doctype="&lt;!DOCTYPE html&gt;"/>
```
Comment (josteinaj) [inline nordic-epub3-dtbook-migrator/src/test/xprocspec/html-to-epub3.step.xprocspec:78]: Any reason why the navigation document had to be first? (just curious)

**Review by josteinaj — COMMENTED** [review]
Looks really great! The only question I have is about whether or not the images references are updated when moving images in and out of th `images/` folder. Also, I'm not able to build this branch, but I don't think that's important as this is work in progress. We'll have to do more practical testing when this is done anyway.

File: dtbook-to-html/src/main/resources/xml/xslt/dtbook-to-epub3.xsl [hunk 3]
```diff
@@ -1032,9 +1031,8 @@
         <xsl:call-template name="f:attrs">
             <xsl:with-param name="all-ids" select="$all-ids" tunnel="yes"/>
         </xsl:call-template>
-        <xsl:attribute name="src" select="concat('images/',@src)"/>
```
[the remark this answers is not in the document]
Comment (bertfrees) [reply dtbook-to-html/src/main/resources/xml/xslt/dtbook-to-epub3.xsl:1035]: This is handled in `px:nordic-dtbook-to-html.step` with a `px:html-update-links` step:

```xml
<px:fileset-filter media-types="image/*" name="dtbook-to-html.step.filter-images">
    <p:input port="source.in-memory">
        <p:pipe step="dtbook-to-html.step.generic" port="result.in-memory"/>
    </p:input>
</px:fileset-filter>
<px:fileset-copy name="dtbook-to-html.step.move-images">
    <p:with-option name="target" select="concat($temp-dir,'images/')"/>
    <p:input port="source.in-memory">
        <p:pipe step="dtbook-to-html.step.filter-images" port="result.in-memory"/>
    </p:input>
</px:fileset-copy>

...

<px:html-update-links>
    <p:input port="mapping">
        <p:pipe step="dtbook-to-html.step.move-images" port="mapping"/>
    </p:input>
</px:html-update-links>
```

File: dtbook-to-html/src/main/resources/xml/xslt/dtbook-to-epub3.xsl [hunk 4]
```diff
@@ -1742,14 +1740,7 @@
     </xsl:template>
     
     <xsl:template match="math:*/@*">
-        <xsl:choose>
-            <xsl:when test="local-name() = 'altimg'">
-                <xsl:attribute name="altimg" select="concat('images/',.)"/>
```
[the remark this answers is not in the document]
Comment (bertfrees) [reply dtbook-to-html/src/main/resources/xml/xslt/dtbook-to-epub3.xsl:1747]: See other comment.

File: html-to-dtbook/src/main/resources/xml/xslt/epub3-to-dtbook.xsl [hunk 5]
```diff
@@ -1484,14 +1466,7 @@
     </xsl:template>
     
     <xsl:template match="math:*/@*">
-        <xsl:choose>
-            <xsl:when test="local-name() = 'altimg'">
-                <xsl:attribute name="altimg" select="replace(.,'^images/','')"/>
```
[the remark this answers is not in the document]
Comment (bertfrees) [reply html-to-dtbook/src/main/resources/xml/xslt/epub3-to-dtbook.xsl:1489]: This is handled in `px:nordic-html-to-dtbook.step` with a `px:dtbook-update-links`.

File: html-to-dtbook/src/main/resources/xml/xslt/epub3-to-dtbook.xsl [hunk 6]
```diff
@@ -244,12 +251,20 @@
                 <xsl:when test="not($headline/preceding-sibling::*[1][f:types(.)='pagebreak']) and $headline/following-sibling::*[1][f:types(.)='pagebreak']">
                     <!-- [tpb126] pagenum must not occur directly after hx unless the hx is preceded by a pagenum -->
                     <xsl:variable name="initial-pagebreak" select="$headline/following-sibling::*[1][f:types(.)='pagebreak']"/>
-                    <xsl:apply-templates select="$initial-pagebreak"/>
```
[the remark this answers is not in the document]
Comment (bertfrees) [reply html-to-dtbook/src/main/resources/xml/xslt/epub3-to-dtbook.xsl:247]: I don't understand. I don't think I removed anything?

File: html-to-dtbook/src/main/resources/xml/xslt/epub3-to-dtbook.xsl [hunk 7]
```diff
@@ -1399,13 +1382,12 @@
         -->
     </xsl:template>
 
-    <!-- <col> is not allowed in nordic DTBook. -->
     <xsl:template match="html:col">
         <xsl:message select="'&lt;col&gt; is not allowed in nordic DTBook.'"/>
```
[the remark this answers is not in the document]
Comment (bertfrees) [reply html-to-dtbook/src/main/resources/xml/xslt/epub3-to-dtbook.xsl:1386]: Ah yes, indeed.

File: nordic-epub3-dtbook-migrator/src/test/xprocspec/html-to-epub3.step.xprocspec (hunk 2 above)
Comment (bertfrees) [reply nordic-epub3-dtbook-migrator/src/test/xprocspec/html-to-epub3.step.xprocspec:78]: I can't immediately pinpoint exactly what caused this change, but it probably happened somewhere in the refactoring of commit 1c0426484223c14398becf7cd0dbe24ccd6ca5c9. Do you really want to know? The fileset XML before/after is equivalent, so...

File: html-to-dtbook/src/main/resources/xml/xslt/epub3-to-dtbook.xsl (hunk 7 above)
[the remark this answers is not in the document]
Comment (bertfrees) [reply html-to-dtbook/src/main/resources/xml/xslt/epub3-to-dtbook.xsl:1386]: I'll fix it.

File: nordic-epub3-dtbook-migrator/src/test/xprocspec/dtbook-to-html.step.xprocspec (hunk 1 above)
Comment (bertfrees) [reply nordic-epub3-dtbook-migrator/src/test/xprocspec/dtbook-to-html.step.xprocspec:35]: Oh, right, forgot about this change. I just checked it again the pending doesn't seem to be needed. Can be changed in https://github.com/nlbdev/nordic-epub3-dtbook-migrator/pull/385.

File: nordic-epub3-dtbook-migrator/src/test/xprocspec/html-to-epub3.step.xprocspec (hunk 2 above)
Comment (josteinaj) [reply nordic-epub3-dtbook-migrator/src/test/xprocspec/html-to-epub3.step.xprocspec:78]: Nah it doesn't really matter.

File: nordic-epub3-dtbook-migrator/src/test/xprocspec/dtbook-to-html.step.xprocspec (hunk 1 above)
Comment (josteinaj) [reply nordic-epub3-dtbook-migrator/src/test/xprocspec/dtbook-to-html.step.xprocspec:35]: Ok, great :+1:

**Comment by bertfrees** [comment]
I have moved the epub3-to-html and html-to-dtbook to pipeline-modules and am now preparing a release of it.

**Comment by josteinaj** [comment]
Ok. Let me know when this can be merged and released, and I'll have a look at it. (I'm going on a vacation now and won't be able to review it until 14th October at the earliest anyway).

**Comment by bertfrees** [comment]
@josteinaj The modules are released now, so I guess this can be merged now. I've updated the branch. Tests still pass.

**Comment by bertfrees** [comment]
@josteinaj Friendly reminder.

**Review by josteinaj — APPROVED** [review]
Looks very nice!

Tests passes, and I'm able to build the docker image. I will test that it works with our production system before releasing.

**Comment by josteinaj** [comment]
@bertfrees I suppose I need a new release of the docker image? The latest seems to be 1.13.4. I've been using 1.12.1 in the Dockerfile.

https://hub.docker.com/r/daisyorg/pipeline-assembly/tags

**Comment by bertfrees** [comment]
Hmm, there hasn't been a Pipeline release yet that includes these modules 1.13.7. But I guess you need it to test it on your production system.

OK, I'll make it now.

**Comment by bertfrees** [comment]
@josteinaj New Docker image is up.

**Comment by josteinaj** [comment]
@bertfrees do I need a descriptor for 1.13.6 to get it working outside of docker?

This does not exist:
https://daisy.github.io/pipeline-assembly/releases/1.13.6

current points to 1.13.5.

**Comment by bertfrees** [comment]
Created.

**Comment by josteinaj** [comment]
Thanks!

**Comment by josteinaj** [comment]
New version is now released and seems to work using both the updater mechanism in Linux and using the Docker image.

**Comment by bertfrees** [comment]
Nice.

**Pull request merged**